[WIP] Submodules rework

batiati opened the pull request:
### This PR:

- Moves the C client to `/clients`.
- Removes the `tigerbeetle` submodule/symlink from the Java and Go client.
- Improves `build.zig` to build and update the clients instead of using sh/bat build scripts.

### Usage:

**Java**: cross-compiles all JNI shared libraries, placing them in the `JAR` folder to be packaged:
```bash
./zig/zig build java_client -Drelease-safe
```

**Go**: cross-compiles all static libraries placing them in the `pkg/native` folder, and also updates the `tb_client.h` header:
```bash
 ./zig/zig build go_client -Drelease-safe
```

### Pending work - WIP

- DotNet client.
- Node client.
- CI
- README and build instructions.

### Open to suggestions:

Should we invoke language-specific commands from `build.zig` ?
For example, the command to build the Java client should be:

**Option 1:** Zig builds only the native code, and language-specific tasks must be invoked separately:

```bash
./zig/zig build java_client -Drelease-safe
cd src/clients/java/src/tigerbeetle-java
mvn package
```
**Option 2:**  Zig builds the whole package. Just calling `./zig/zig build java_client -Drelease-safe` also invokes Java/Maven specific commands.

```bash
./zig/zig build java_client -Drelease-safe
```

The same logic can be applied to running tests, executing samples, benchmarks, etc.

**Option 3:** Rely on language-specif toolchains for building the clients. Just calling `mvn package` also invokes `zig build`.

```bash
cd src/clients/java/src/tigerbeetle-java
mvn package
```

sentientwaffle committed:
StateMachine: Unit test tables

Refactor the state machine unit tests to a more compact form.

Rationale: The state machine tests were so verbose that they were hard to skim through or verify manually. This new layout makes it easy to check for e.g. missing test cases, and makes the organization more obvious (descending error precedence). In the process of this I found several missing tests (not missing codes, just multiple cases that could hit the code).

sentientwaffle committed:
Config: Remove unused configs

sentientwaffle committed:
Config: Distinguish config from constants

There are (roughly) three categories of constants in `config.zig`:
- `ConfigCluster` constants can be modified when the cluster is first configured, but must never change after that.
- `ConfigProcess` can vary per-replica/client, and can be changed between restarts.
- And various constants derived from the above (plus a few that just have fixed values, like `sector_size`).

I strongly considered threading a `constants` type through all of the type constructors from `main.zig`/`simulator.zig` on down. But this adds quite a lot of "noise" to the code, since there are many functions/constants/types that aren't generic already.

The main goal is to allow the VOPR and other fuzzers to test different configurations — this is still achievable with to "global" `config.zig`.

The downsides of the "global" `config.zig`:
- Only one config can be run per process. e.g. the VOPR cannot test multiple configs simultaneously.
- If a unit test needs to target a specific configuration, it needs to specifically factor out the config. (See `SuperBlockQuorums` for an example).

sentientwaffle committed:
StateMachine: Test strings

sentientwaffle committed:
StateMachine: Test public interface

sentientwaffle committed:
Test: Clean up table parser

sentientwaffle committed:
StateMachine: Test rollback within a commit

sentientwaffle committed:
Config: Remove config.Environment

sentientwaffle committed:
Config: Run test with config_test_min

sentientwaffle committed:
Config: Minimize test config

sentientwaffle committed:
LSM: Allow zero-sized value cache

This also moves ownership of the value cache from the groove(s) to the tree, just to clean things up.

sentientwaffle committed:
Config: Disable transfer value cache by default

Also increase the sizer of the account cache, relative to the posted-transfer cache. (For reasons explained in `config.zig`).

sentientwaffle committed:
Config: Get base config from "zig build"

By default:
- When running tests, use `test_min`.
- For fuzzers, use `test_min`.
- Otherwise use `development`.

sentientwaffle committed:
Config: Rename io_depth_* → journal_iops_*_max

These config optional only apply to the journal. The grid's IOPS limits are presently hard-coded (see https://github.com/tigerbeetledb/tigerbeetle/issues/295).

sentientwaffle committed:
Config: Manually export "log"

To match `log_level`.

sentientwaffle committed:
Config: Simplify build options

sentientwaffle committed:
Config: Spread out assertions

Config constraints are borderline documentation, so keep them close to their respective declarations.

sentientwaffle committed:
Test/Table: Clarify code

sentientwaffle committed:
Config: Rename to constants

This only renames the file — it is still referred to `config`.
(I don't want to trigger a zillion git conflicts...)

sentientwaffle committed:
Config: Fix node client tracer backend

sentientwaffle committed:
Test/Table: Simplify tests

Unfortunately the MacOS CI test segfaults when using a BoundedArray.
Once we upgrade Zig we should be able to just use slices anyway, so until then use ArrayList as a stopgap.

batiati committed:
Submodule rework

- Move the C client to the /clients folder.
- Remove tigerbeetle submodule from the Java client.
- Adjust build and paths

batiati committed:
Submodule rework

- Move the C client to the /clients folder.
- Remove tigerbeetle submodule from the Java client.
- Adjust build and paths

sentientwaffle committed:
CLI: Allow cache=0

batiati committed:
Go client build

- Replaces the script `rebuild_binaries.sh` with a zig build:
`zig build go_client -Drelease-safe`

- Updates the `tb_client.h` header

batiati committed:
Go client fixes

batiati committed:
Zig fmt

sentientwaffle committed:
StateMachine: Disable tests on MacOS only

Disabled because of spurious failure in MacOS CI.
Most likely due to a compiler bug that will be fixed when we upgrade to 0.10.

(Example run: https://github.com/tigerbeetledb/tigerbeetle/actions/runs/3587237036/jobs/6037355308)

sentientwaffle committed:
Fuzz,CI: Add --events-max

Allow a runtime argument to limit the number of fuzzer repetitions.

CI canary runs can be fast, but for actual fuzzing we can still use a high default events-max.

(The motivation for this was a (successful) ManifestLog fuzzer run that took >5 hours.)

batiati committed:
Java client: Add support for musl-based Linux distros and fix Mac OS detection (#268)

* Java client: Add support for musl-based distros

- Cross-compiles for linux using glibc and musl: The JVM can only load the JNI lib that matches its libc, so we must distribute both versions inside the .jar package.

- Simplifies the cross-compilation naming process, changing the lib subdirectory's names to match the zig triple `arch-os-abi`.

- Detects musl-based distros during runtime by inspecting the process's loaded libraries.

- Minor adjustments for impacted test cases.

- Also, closes https://github.com/tigerbeetledb/tigerbeetle-java/pull/21

* Add build options workaround

- Added a workaround for the required build options introduced by the tracing API #291
This should be removed once the client is built using the root build.zig file

* Zig fmt

batiati committed:
Java CI tests (#284)

* Java CI tests

* Update GitHub Maven registry path

* CI: yaml array style

Use yaml array style for large arrays in favor of the JS style.

sentientwaffle committed:
Build: Dedup fuzzer build steps

sentientwaffle committed:
CI: Run tree fuzzer canary

As of https://github.com/tigerbeetledb/tigerbeetle/pull/301 the tree fuzzer can reliably pass. Use a low `events-max` so that CI is fast.

sentientwaffle committed:
CI: Label weekly scheduled build

sentientwaffle committed:
Main: Group config values (#308)

Group config output into "build", "cluster", and "process".

sentientwaffle committed:
Config: Rename "config" identifier to "constants"

sentientwaffle committed:
VOPR: Use configs.test_min

Missed this fuzzer when I changed how configs are injected by fuzzers.

sentientwaffle committed:
Clients/Node: Bump version

kprotty commented on build.zig:
```diff
@@ -404,3 +406,110 @@ fn link_tracer_backend(
+        lib.pie = true;
+        lib.bundle_compiler_rt = true;
+
+        if (cross_target.os_tag.? != .windows) {
+            lib.linkLibC();
+        }
```
We can always linkLibC() on windows now (`builtin.link_libc` is used to [know when to export](https://github.com/tigerbeetledb/tigerbeetle/blob/main/src/c/tb_client.zig#L57-L65) the tb_client functions)

batiati replied:
Thanks for the tip! That's why the Java CI is failing right now on Windows!

kprotty reviewed (commented):
Nice work! Really cleans up the build process for the clients. One last tidbit for C header generation on windows.

sentientwaffle committed:
Clients/Java: Workaround tigerbeetle_build_options (#312)

kprotty commented:
Need to rebase on main + address the CI issues though

batiati committed:
Fix Java client on Windows

- Link libC to export functions.

- Fix native lib path to match zig triple os-arch-abi.

sentientwaffle committed:
CI: Deduplicate PR workflows (#306)

batiati committed:
Rebase adjustments

batiati committed:
Dotnet client submodule removal

batiati committed:
CI and scripts adjustments

- Fixes Go CI
- Updates Java and Dotnet scripts

batiati commented:
superseded by #320
